### ISSUE TITLE
Setting VNC password breaks web VNC viewer

### ISSUE THREAD

**Issue by brandonkal** [cockpit-project/cockpit-machines#1405]
This doesn't seem to be completely fixed in cockpit-machines 304-1~bpo22.04.1 (on Ubuntu LTS 22.04 host)
To reproduce:
1. Changing qemu.conf file to include a password and restarting libvirtd.
2. This breaks existing machines VNC webviewer
3. Using virsh edit domain I can add a passwd to the vnc XML line but when cockpit starts the VM that is ignored.
4. Web VNC does display but flashes in and out. This doesn't seem to be an issue when connecting to the VM via Apple VNC viewer with the password.

Also broken is setting qemu.conf to listen on 0.0.0.0 for VNC is not respected by cockpit, it is always 127.0.0.1.

**Comment by jelly** [comment]
That doens't sound ideal, we have an issue to make this configurable which should also handle the password scenario then. https://github.com/cockpit-project/cockpit-machines/issues/553 

However: Cockpit should not change qemu.conf and I am not sure if we want to parse it.

**Comment by mvollmer** [comment]
> Also broken is setting qemu.conf to listen on 0.0.0.0 for VNC is not respected by cockpit, it is always 127.0.0.1.

Hmm, can you elaborate of what you mean with "respected by Cockpit"? Cockpit displays the listening address in the "Desktop viewer" tab, but does not for example influence what address qemu is actually listening on when started (that is controlled by libvirt alone). Do you see the wrong listening address in the "Desktop viewer" tab?

**Comment by mvollmer** [comment]
> - Changing qemu.conf file to include a password and restarting libvirtd.
> - This breaks existing machines VNC webviewer

Yes, that's a bug. Cockpit never prompts for a VNC password (see below), but instead retrieves it from libvirt and passes it silently to the in-page VNC viewer. Cockpit does however only look into the machine definition for the password, not into qemu.conf.

> - Using virsh edit domain I can add a passwd to the vnc XML line but when cockpit starts the VM that is ignored.

I think the password is not ignored; it is still required to connect to the VNC server with viewers like GNOME Connections or remote-viewer. Cockpit itself however is sneaky and looks into the XML definition for the password and just uses it. This is why the in-page viewer now works.

We should at least change Cockpit to prompt for a password if the viewer requests one and Cockpit hasn't found one itself.

We could also make Cockpit look into qemu.conf to find the password. Or, we could change it to never look for passwords.

**Comment by mvollmer** [comment]
> > Also broken is setting qemu.conf to listen on 0.0.0.0 for VNC is not respected by cockpit, it is always 127.0.0.1.
> 
> Hmm, can you elaborate of what you mean with "respected by Cockpit"? Cockpit displays the listening address in the "Desktop viewer" tab, but does not for example influence what address qemu is actually listening on when started (that is controlled by libvirt alone). Do you see the wrong listening address in the "Desktop viewer" tab?

This was very likely fixed by https://github.com/cockpit-project/cockpit-machines/pull/2287

**Comment by mvollmer** [comment]
> > * Changing qemu.conf file to include a password and restarting libvirtd.
> > * This breaks existing machines VNC webviewer
> 
> Yes, that's a bug. Cockpit never prompts for a VNC password (see below), but instead retrieves it from libvirt and passes it silently to the in-page VNC viewer. Cockpit does however only look into the machine definition for the password, not into qemu.conf.

This was fixed by https://github.com/cockpit-project/cockpit-machines/pull/2291.

**Comment by mvollmer** [comment]
All fixed, I think!
